# Part 1. Equation of a Slime

How many late days are you using for this assignment?

0

In [269]:
# Imports section
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import numpy as np




## 1. Loading the dataset

In [270]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
dataset1 = pd.read_csv('science_data_large.csv')

# Print the data
print(dataset1.head(15))


    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04


# Display a summary of the table information (number of datapoints, etc.)

There are 1000 entries with 3 columns: Temperature °C, Mols KCL, Size nm^3. All columns have non-null values. 

## 2. Splitting the dataset

In [271]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = dataset1[['Temperature °C', 'Mols KCL']]  
y = dataset1['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)


## 3. Perform a Linear Regression

In [272]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sampleDatapoint = X_test.iloc[0].values.reshape(1, -1)
predictedOutput = model.predict(sampleDatapoint)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score = model.score(X_test, y_test)
print(score)
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_
print(coefficients)
print(intercept)



0.8552472077276096
[ 866.14641337 1032.69506649]
-409391.4795834081


/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = 866.1464 \cdot x_1 + 1032.6951 \cdot x_2 - 409391.4796$

Report on score and explain meaning:

The score we reported was the R^2 score which shows how well the independent variables accurately predict the variance. An R^2 score of 1 means the model perfectly predicts the actual results. An R^2 score of 0 means the model does not explain any variance in the data. We got an R^2 score of 0.8552472077276096 so the model's prediction is pretty accurate as it is very close to 1. 

## 4. Use Cross Validation

In [273]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42


cv = KFold(n_splits=5, shuffle=True, random_state=42)
score1 = cross_val_score(model, X, y, cv=cv, scoring='r2')

# Report on their finding and their significance
print(score1)

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here:

Mean R^2 score: (0.86151889 + 0.82742341 + 0.87195173 + 0.88166206 + 0.85609101)/5 = 0.85972942 
This represents the average performance of the model across different train-test splits which is 0.85972942, so it explain about 86% of the variance.  As previously mentioned, the closer an R^2 score is to 1, the better performing it is. Thus, the model is performing well. The scores across the five folds are consistent, so the model generalizes well and isn't overly sensitive to different data splits. 


## 5. Using Polynomial Regression

In [274]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)
modelPoly = LinearRegression()
modelPoly.fit(X_train, y_train)

scorePoly = modelPoly.score(X_test, y_test)
cv = KFold(n_splits=5, shuffle=True, random_state=42)
polyScores = cross_val_score(modelPoly, X_poly, y, cv=cv, scoring='r2')

coefficientsPoly = modelPoly.coef_
interceptPoly = modelPoly.intercept_

print(polyScores)
print(coefficientsPoly)
print(interceptPoly)

# Report on the metrics and output the resultant equation as you did in Part 3.

[1. 1. 1. 1. 1.]
[ 0.00000000e+00  1.20000000e+01 -1.27198206e-07  1.26473309e-11
  2.00000000e+00  2.85714287e-02]
2.047908492386341e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$ h(x) = (2.0479 \times 10^{-5}) + (0.0)x + (12.0)x^2 + (-1.27198 \times 10^{-7})x^3 + (1.26473 \times 10^{-11})x^4 + (2.0)x^5 + (0.02857)x^6 $


Report on the score and interpret:

The R^2 value is 1.0. As previously mentioned, the closer the score is to 1, the better the prediction the model gives. A score of 1.0 means the model perfectly fits the training and test data. This is rare in real world applications, but the dataset used for this is synthetic, so this high of a score is expected.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [275]:
# Load the dataset. Then train and evaluate the classification models.
dataset2 = pd.read_csv('ckd_feature_subset.csv')

#prepare target
X = dataset2.drop('Target_ckd', axis=1)
y = dataset2['Target_ckd']

#don't need to handle missing values bc there are none

#Traintest split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#Logistic Regression model 
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)
logisticPrediction = logistic_model.predict(X_test)

#SVC model
SVC_model = SVC(decision_function_shape='ovo')
SVC_model.fit(X_train, y_train)
SVC_prediction = SVC_model.predict(X_test)

#knearest model
kNearest_model = KNeighborsClassifier()
kNearest_model.fit(X_train, y_train)
kNearest_prediction = kNearest_model.predict(X_test)

#neural networks model
nn_model = MLPClassifier(
    max_iter=1000,
    activation='relu',
    hidden_layer_sizes=100,
    random_state=42
)
nn_model.fit(X_train, y_train)
nn_prediction = nn_model.predict(X_test)

## Results and Conclusion for Classification Experiments

In [276]:
#5-fold cross validation 

cv_lr = cross_val_score(logistic_model, X, y, cv=5)

cv_svc = cross_val_score(SVC_model, X, y, cv=5)

cv_knn = cross_val_score(kNearest_model, X, y, cv=5)

cv_nn = cross_val_score(nn_model, X, y, cv=5)

cv_results = [
    {"Model": "Logistic Regression", "Average Accuracy": cv_lr.mean(), "Standard Deviation": cv_lr.std()},
    {"Model": "Support Vector Machine", "Average Accuracy": cv_svc.mean(), "Standard Deviation": cv_svc.std()},
    {"Model": "k-Nearest Neighbors", "Average Accuracy": cv_knn.mean(), "Standard Deviation": cv_knn.std()},
    {"Model": "Neural Network", "Average Accuracy": cv_nn.mean(), "Standard Deviation": cv_nn.std()}
]

# Convert to a DataFrame
results_df = pd.DataFrame(cv_results)

# Display the table
print(results_df)



                    Model  Average Accuracy  Standard Deviation
0     Logistic Regression          0.856129            0.054236
1  Support Vector Machine          0.928172            0.059993
2     k-Nearest Neighbors          0.928172            0.031893
3          Neural Network          0.941290            0.024040


The best performing model is neural network with average accuracy of 0.941 which is the highest across all models and a standard deviation of 0.024 which is the lowest among all models. Thus, as it has the highest average accuracy and lowest standard deviation it is the most relaible model among the four. 
Next, SVC and k-Nearest Neighbors have the same accuracy with an average accuracy of 0.928. However, k-Nearest Neighbors had a lower standard deviation so it's more consistent than SVC across the 5 folds. Finally, the worst performing model is logistic regression with an average accuracy of 0.856 and a standard deviation of 0.0542. 


## Results and Conclusion for Neural Network Experiments

In [277]:
configurations = {
    'Config 1': MLPClassifier(hidden_layer_sizes=(10,), activation='relu', random_state=42, max_iter=5000),
    'Config 2': MLPClassifier(hidden_layer_sizes=(50,), activation='tanh', random_state=42, max_iter=5000),
    'Config 3': MLPClassifier(hidden_layer_sizes=(100, 50), activation='logistic', random_state=42, max_iter=5000)
}

# Store results for configurations
nn_results = {}

for config_name, model in configurations.items():
    scores = cross_val_score(model, X_train, y_train, cv=5)
    nn_results[config_name] = {
        'average_accuracy': np.mean(scores),
        'std_accuracy': np.std(scores)
    }

# Convert to a DataFrame 
nn_results_df = pd.DataFrame(nn_results).T
print(nn_results_df)

          average_accuracy  std_accuracy
Config 1          0.926000      0.047593
Config 2          0.934667      0.019362
Config 3          0.738000      0.014697


Config 2, which had one hidden layer with 50 neurons and 'tanh' activation is the best performing as it has a average accuracy of 0.934. It had the second lowest standard deviation wit 0.019. This shows that a moderately sized network with a smooth activation function like tanh was able to capture important patterns in the CKD dataset without overfitting or underfitting. 

Config 1, which used a smaller architecture with 10 neurons in one hidden layer and the relu activation, performed second best with an average accuracy of 0.926. Config 1 had the highest standard deviation among the three with 0.047. 

Config 3 had a deeper architecture with two hidden layers (100 and 50 neurons) and used the logistic activation function. Even though it was the most complex, it performed the worst with an average accuracy of 0.738. This result may be due to convergence issues or overfitting. Although it did have the lowest standard deviation with 0.014.

This shows that a more complex model doesn't necessarily lead to better performance. In this case, the best performing model was a balanced, single-layer network, so tuning of activation functions and network depth is more important than just increasing size or complexity. 


